### PR TITLE
Remove (most of) the code intended for python 2

### PR DESCRIPTION
**Pull request by jhaigh0**
### Description of work

Remove code which is taking precautions wrt python 2. Release 4.2.0 was the final release to support python 2, so I think there's little risk in doing this.
Also updated how `simpleapi.py` was handling function signatures as part of this.

I've only searched for code by looking for comments mentioning python 2 so I'm sure there are more places out there.

There are a couple of cpp boost instances which look like they could be changed, but this is supposed to be a quick maintenance task and they look like they'd be more complex problems. Let me know if you disagree.
https://github.com/mantidproject/mantid/blob/main/Framework/PythonInterface/core/inc/MantidPythonInterface/core/PythonObjectInstantiator.h#L80-L87
https://github.com/mantidproject/mantid/blob/main/Framework/PythonInterface/mantid/kernel/src/Registry/PropertyWithValueFactory.cpp#L54-L61

There are some comments `implement with ContextDecorator after python2 is deprecated)` which have been deleted. They are currently using `@contextmanager` which is based on `ContextDecorator` anyway 

*There is no associated issue.*

### To test:

- ci tests pass
- feel free to search for more code that I've missed

*This does not require release notes* because **no user impact**

---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

### CODE REVIEW

File: Framework/PythonInterface/mantid/simpleapi.py [hunk 1]
```diff
@@ -527,7 +525,7 @@ def CutMD(*args, **kwargs):  # noqa: C901
         return out_names[0]
 
 
-_replace_signature(CutMD, ("\bInputWorkspace", "**kwargs"))
+CutMD.__signature__ = LazyFunctionSignature(alg_name="CutMD", include_self=False)
```
Comment (jclarkeSTFC) [inline Framework/PythonInterface/mantid/simpleapi.py:528]: Do you know why some algorithms need special treatment? E.g. `CutMD` and `RenameWorkspace`

File: Framework/PythonInterface/mantid/api/src/Exports/SpectrumInfoPythonIterator.cpp [hunk 2]
```diff
@@ -7,7 +7,6 @@
 #include "MantidPythonInterface/api/SpectrumInfoPythonIterator.h"
 
 #include <boost/python/class.hpp>
-#include <boost/python/copy_const_reference.hpp>
 #include <boost/python/iterator.hpp>
 #include <boost/python/module.hpp>
```
Comment (jclarkeSTFC) [inline Framework/PythonInterface/mantid/api/src/Exports/SpectrumInfoPythonIterator.cpp:11]:
```suggestion
#include <boost/python/module.hpp>
#include <boost/python/reference_existing_object.hpp>
```
I know it compiles as it is, but seems safer to include the header. I can't actually see how that header comes into this file currently, must come from somewhere.

**Review by jclarkeSTFC — CHANGES_REQUESTED** [review]
Looks good, just a couple of minor points. I also fired up Workbench (no errors or warnings) and tried a couple of the algorithms that have changes, worked fine, as far as I could tell.

With the changes to `simpleapi` might be worth considering a package build.

**Review by jclarkeSTFC — DISMISSED** [review]

**Review by jclarkeSTFC — DISMISSED** [review]

File: buildconfig/CMake/Packaging/AddPythonPath.py.in [hunk 3]
```diff
@@ -46,27 +46,8 @@ for directory in pathdirs:
...
-                except ImportError:
-                    die("Did not fix import error")
-                print("       {}  ... adding to mantid.pth".format(" " * len(platform_path)))
-                pathdirs.append(platform_path)
-                found_dlfcn = True
-        if not found_dlfcn:  # missing path wasn't found
-            die()
-    else:
-        die("Can't import mantid: {}".format(e))
+    die("Can't import mantid: {}".format(e))
```
Comment (thomashampson) [inline buildconfig/CMake/Packaging/AddPythonPath.py.in:49]:
```suggestion
    die(f"Can't import mantid: {e}")
```

File: scripts/SANS/isis_instrument.py [hunk 4]
```diff
@@ -1340,7 +1340,7 @@ def _get_const_num(self, log_data, log_name):
             # return the log value if it stored as a single number
             return float(log_data.getLogData(log_name).value)
         except TypeError:
-            # Python 2.4 doesn't have datetime.strptime...
+
```
[the remark this answers is not in the document]
Comment (thomashampson) [reply scripts/SANS/isis_instrument.py:1343]: Can you comment on how this is resolved?

**Review by thomashampson — CHANGES_REQUESTED** [review]
Looking good, just a couple of comments

**Review by thomashampson — APPROVED** [review]

**Pull request merged**